### DISCUSSION TITLE
running on strix halo and other shenanigans

### DISCUSSION THREAD

**Discussion by LokiMetaSmith** [karpathy/nanochat#363] [Show and tell]
so, I think it's all working, but then I started adding random features, so now there's a feature for auto tuning parameters, a server for running automated training, a few experimental training methods that looked interesting, and a run.sh script for running python (because the virtual env was wonky to run python scripts outside of the venv) 

https://github.com/LokiMetaSmith/nanochat

I guess what I'm asking is, I'm fine spinning off into fork land, but is there interest in merging some of these features as a pull request back into the main project

here's an ai summary: 

1. Infrastructure: AMD / ROCm & Strix Halo Support
Why: Brings nanochat to a much wider range of hardware, including consumer APUs (Strix Halo) and AMD GPUs, which is consistent with the goal of "democratizing" LLM training.

Implementation:
pyproject.toml: Dedicated amd dependency group pointing to specific ROCm wheels.
nanochat/common.py: Auto-detection of ROCm, fallback to gloo backend (since NCCL is unstable on APUs), and HSA environment overrides.
scripts/base_train.py: Logic to enable TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL and handle ld.lld paths for Triton compilation.
tinyrun.sh / tunerun.sh: Integrated setup logic to uninstall conflicting triton and install pytorch-triton-rocm.
2. Research: Online Curriculum Learning (Infovore)
Why: A sophisticated, "free" improvement to training efficiency that dynamically weights data based on the model's current knowledge boundary.

Implementation:
nanochat/infovore.py: Implements the Infovore agent and Novelty-Relation Quotient (NRQ) metric.
scripts/base_train.py: Clean integration via --use_infovore flag.
3. Research: Nested Momentum Optimizer
Why: Implements cutting-edge optimization research (Behrouz et al., 2025) which may offer better convergence or stability than standard AdamW/Muon.

Implementation:
nanochat/nl_opt.py: NestedMomentum and its distributed ZeRO-2 variant DistNestedMomentum.
scripts/base_train.py: Support for matrix_optimizer_backend="nested_momentum".
4. Tooling: tinyrun.sh & tunerun.sh
Why: Drastically lowers the barrier to entry. tinyrun.sh allows running on a single consumer GPU (RTX 3080), whereas upstream focuses on 8xH100 (speedrun.sh).

Implementation:
tinyrun.sh: Automated "single-click" setup for small scale.
tunerun.sh: Automated system performance tuning (grid search over batch sizes/compilation).
5. Tooling: rustbpe (Likely Fork Feature)
Why: A middle-ground tokenizer: faster than Python minbpe but simpler than HuggingFace tokenizers.

Implementation:
rustbpe/: Rust source code.
Builds via maturin in the setup scripts.

Recommendation: The AMD Support and Tinyrun scripts are the most "product-ready" and aligned with the project's goal of accessibility. Infovore and NestedMomentum are valuable "Advanced/Experimental" features.

**Comment by gbossi** [comment]
I have extended the installation script to support both the gfx1151 and gfx120x AMD architectures (9060xt, 9070xt r9700), but I have split the installation part from the model train and test scripts. This is the fork I am working:

[https://github.com/gbossi/nanochat](url)

In this fork the installation script is called setup_venv.sh and it will auto-detect your own GPU and if it is supported it will configure and install the virtual environment according to the hardware installed.

**Reply by LokiMetaSmith** [reply]
I've incorporated those changes in my fork. 
would be curious if they run on different hardware now.

**Comment by yurivilmanis** [comment]
I'd be interested in the Strix Halo ROCm support, so I guess that makes one? I notice there's a pull request for ROCm support now though, so that may cover it [(Add support for ROCm backend in speedrun script)](https://github.com/karpathy/nanochat/pull/409)

The Strix Halo mini PCs with 128GB shared memory seem to be a relatively low-cost, low-power solution with enough memory to handle training small models. I've been using mine mainly for running LLM sessionse, but if you already have a main PC for doing other things, and a "day job" that means your tinkering is done mainly on weekends, I suspect it would be great for doing a fire-and-forget training run over a week or two, just chugging along in the background... definitely democratising (albeit relatively slow) training of models with more parameters or longer context than the typical 'gaming' GPU with 16 or 24GB VRAM.

**Comment by khimaros** [comment]
here's a minimal patch which works on strix halo. it does not change any default behavior except for more strict bash settings (`set -euo pipefail`).

however, performance is quite terrible. i'm seeing ~3500t/s and i estimate that the `base_train` script will take about 185 days to run at this pace :)

that said, hopefully it is in a reasonable shape to upstream since it is a minimal delta against the existing repository.

## instructions

save the patch somewhere, apply it, and then execute `COMPUTE_BACKEND=rocm NPROC_PER_NODE=1 bash -x ./runs/speedrun.sh`

## patch

```diff
diff --git a/.python-version b/.python-version
index c8cfe39..e4fba21 100644
--- a/.python-version
+++ b/.python-version
@@ -1 +1 @@
-3.10
+3.12
diff --git a/pyproject.toml b/pyproject.toml
index f3cd8d7..03abf6b 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -45,6 +45,20 @@ python_functions = ["test_*"]
 torch = [
     { index = "pytorch-cpu", extra = "cpu" },
     { index = "pytorch-cu128", extra = "gpu" },
+    { index = "pytorch-rocm-gfx1151", extra = "rocm-gfx1151" },
+]
+
+rocm = [
+    { index = "pytorch-rocm-gfx1151", extra = "rocm-gfx1151" },
+]
+pytorch-triton-rocm = [
+    { index = "pytorch-rocm-gfx1151", extra = "rocm-gfx1151" },
+]
+rocm-sdk-core = [
+    { index = "pytorch-rocm-gfx1151", extra = "rocm-gfx1151" },
+]
+rocm-sdk-libraries-gfx1151 = [
+    { index = "pytorch-rocm-gfx1151" },
 ]
 
 [[tool.uv.index]]
@@ -57,6 +71,12 @@ name = "pytorch-cu128"
 url = "https://download.pytorch.org/whl/cu128"
 explicit = true
 
+[[tool.uv.index]]
+name = "pytorch-rocm-gfx1151"
+#url = "https://rocm.nightlies.amd.com/v2-staging/gfx1151"
+url = "https://repo.amd.com/rocm/whl/gfx1151"
+explicit = true
+
 [project.optional-dependencies]
 cpu = [
     "torch>=2.9.1",
@@ -64,11 +84,21 @@ cpu = [
 gpu = [
     "torch>=2.9.1",
 ]
+rocm-gfx1151 = [
+    "torch>=2.9.1,<2.10.0",
+    "torchvision",
+    "rocm",
+    "pytorch-triton-rocm",
+    "rocm-sdk-core",
+    "rocm-sdk-libraries-gfx1151",
+]
 
 [tool.uv]
+prerelease = "allow"
 conflicts = [
     [
         { extra = "cpu" },
         { extra = "gpu" },
+        { extra = "rocm-gfx1151" },
     ],
 ]
diff --git a/runs/speedrun.sh b/runs/speedrun.sh
index 8fff564..ae67bbc 100644
--- a/runs/speedrun.sh
+++ b/runs/speedrun.sh
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -euo pipefail
+
 # This script is the "Best ChatGPT clone that $100 can buy",
 # It is designed to run in ~4 hours on 8XH100 node at $3/GPU/hour.
 
@@ -17,13 +19,22 @@ mkdir -p $NANOCHAT_BASE_DIR
 
 # -----------------------------------------------------------------------------
 # Python venv setup with uv
+COMPUTE_BACKEND=${COMPUTE_BACKEND:-"cuda"}
+
+PYTHON_VENV_EXTRA="gpu"
+if [[ "$COMPUTE_BACKEND" = "rocm" ]]; then
+  export TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1
+  PYTHON_VENV_EXTRA="rocm-gfx1151"
+elif [[ "$COMPUTE_BACKEND" = "cpu" ]]; then
+  PYTHON_VENV_EXTRA="cpu"
+fi
 
 # install uv (if not already installed)
 command -v uv &> /dev/null || curl -LsSf https://astral.sh/uv/install.sh | sh
 # create a .venv local virtual environment (if it doesn't exist)
 [ -d ".venv" ] || uv venv
 # install the repo dependencies
-uv sync --extra gpu
+uv sync --extra ${PYTHON_VENV_EXTRA}
 # activate venv so that `python` uses the project's venv instead of system python
 source .venv/bin/activate
 
@@ -34,10 +45,8 @@ source .venv/bin/activate
 #    `wandb login`
 # 2) Set the WANDB_RUN environment variable when running this script, e.g.:
 #    `WANDB_RUN=d26 bash speedrun.sh`
-if [ -z "$WANDB_RUN" ]; then
-    # by default use "dummy" : it's handled as a special case, skips logging to wandb
-    WANDB_RUN=dummy
-fi
+# by default use "dummy" : it's handled as a special case, skips logging to wandb
+WANDB_RUN=${WANDB_RUN:-"dummy"}
 
 # -----------------------------------------------------------------------------
 # During the course of the run, we will be writing markdown reports to the report/
@@ -78,14 +87,23 @@ echo "Waiting for dataset download to complete..."
 wait $DATASET_DOWNLOAD_PID
 
 # Number of processes/GPUs to use
-NPROC_PER_NODE=8
+NPROC_PER_NODE=${NPROC_PER_NODE:-8}
+
+run() {
+  script="$1"; shift
+  if (( NPROC_PER_NODE > 1 )); then
+    torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m "$script" -- "$@"
+  else
+    python -m "$script" "$@"
+  fi
+}
 
 # pretrain the d20 model
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.base_train -- --depth=20 --target-param-data-ratio=20 --run=$WANDB_RUN
+run scripts.base_train --depth=20 --target-param-data-ratio=20 --run=$WANDB_RUN
 # evaluate the model on a larger chunk of train/val data and draw some samples
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.base_loss
+run scripts.base_loss
 # evaluate the model on CORE tasks
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.base_eval
+run scripts.base_eval
 
 # -----------------------------------------------------------------------------
 # Midtraining (teach the model conversation special tokens, tool use, multiple choice)
@@ -95,15 +113,15 @@ torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.base_eval
 curl -L -o $NANOCHAT_BASE_DIR/identity_conversations.jsonl https://karpathy-public.s3.us-west-2.amazonaws.com/identity_conversations.jsonl
 
 # run midtraining and eval the model
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.mid_train -- --run=$WANDB_RUN
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.chat_eval -- -i mid
+run scripts.mid_train --run=$WANDB_RUN
+run scripts.chat_eval -i mid
 
 # -----------------------------------------------------------------------------
 # Supervised Finetuning (domain adaptation to each sequence all by itself per row)
 
 # train sft and re-eval right away (should see a small bump)
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.chat_sft -- --run=$WANDB_RUN
-torchrun --standalone --nproc_per_node=$NPROC_PER_NODE -m scripts.chat_eval -- -i sft
+run scripts.chat_sft --run=$WANDB_RUN
+run scripts.chat_eval -i sft
 
 # chat with the model over CLI! Leave out the -p to chat interactively
 # python -m scripts.chat_cli -p "Why is the sky blue?"
```
